private/buf/cmd/buf/command/beta/studioagent/studioagent.go

```diff
@@ -0,0 +1,241 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package studioagent
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/bufbuild/buf/private/buf/bufcli"
+	"github.com/bufbuild/buf/private/bufpkg/bufstudioagent"
+	"github.com/bufbuild/buf/private/pkg/app/appcmd"
+	"github.com/bufbuild/buf/private/pkg/app/appflag"
+	"github.com/bufbuild/buf/private/pkg/stringutil"
+	"github.com/bufbuild/buf/private/pkg/thread"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
+)
+
+const (
+	// defaultShutdownTimeout is the default shutdown timeout.
+	defaultShutdownTimeout = 10 * time.Second
+	// defaultReadHeaderTimeout is the default read header timeout.
+	defaultReadHeaderTimeout = 30 * time.Second
+	// defaultIdleTimeout is the amount of time an HTTP/2 connection can be idle.
+	defaultIdleTimeout = 3 * time.Minute
+)
+
+const (
+	portFlagName              = "port"
+	disallowedHeadersFlagName = "disallowed-header"
+	forwardHeadersFlagName    = "forward-header"
+	caCertFlagName            = "ca-cert"
+	clientCertFlagName        = "client-cert"
+	clientKeyFlagName         = "client-key"
+	serverCertFlagName        = "server-cert"
+	serverKeyFlagName         = "server-key"
+)
+
+// NewCommand returns a new Command.
+func NewCommand(
+	name string,
+	builder appflag.Builder,
+) *appcmd.Command {
+	flags := newFlags()
+	return &appcmd.Command{
+		Use:   name + " <origin>",
+		Short: "Run an HTTP(S) server as the studio agent with the origin be set as the allowed origin for CORS options.",
+		Args:  cobra.ExactArgs(1),
+		Run: builder.NewRunFunc(
+			func(ctx context.Context, container appflag.Container) error {
+				return run(ctx, container, flags)
+			},
+			bufcli.NewErrorInterceptor(),
+		),
+		BindFlags: flags.Bind,
+	}
+}
+
+type flags struct {
+	Port              string
+	DisallowedHeaders []string
+	ForwardHeaders    map[string]string
+	CACert            string
+	ClientCert        string
+	ClientKey         string
+	ServerCert        string
+	ServerKey         string
+}
+
+func newFlags() *flags {
+	return &flags{}
+}
+
+func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	flagSet.StringVar(
+		&f.Port,
+		portFlagName,
+		"8080",
+		"The port to be exposed to accept HTTP requests.",
+	)
+	flagSet.StringSliceVar(
+		&f.DisallowedHeaders,
+		disallowedHeadersFlagName,
+		nil,
+		`The headers to be disallowed via the agent to the target server. Multiple headers are appended if specified multiple times.`,
```

Comment:
nit: "The headers disallowed via the agent to the target server. Multiple headers are appended if specified multiple times."